### PR TITLE
Clarify error message for one cause of invalid argument prfix `!`

### PR DESCRIPTION
**Pull request by muzimuzhi**
Implements #1198

Rollback needed?

**READ ME FIRST**: Please understand that in most cases we will not be able to merge a pull request because there are a lot of internal activities needed when updating the LaTeX2e sources. If you have a code suggestion please discuss it with the team first.

*Pull requests in this repository are intended for LaTeX Team members only.*

# Internal housekeeping

## Status of pull request

- Feedback wanted 
- Under development

## Checklist of required changes before merge will be approved
- [ ] Test file(s) added
- [x] Version and date string updated in changed source files
- [x] Relevant `\changes` entries in source included
- [x] Relevant `changes.txt` updated
- [ ] Rollback provided (if necessary)?
- [ ] `ltnewsX.tex` (and/or `latexchanges.tex`) updated

### CODE REVIEW

File: base/changes.txt [hunk 1]
```diff
@@ -6,6 +6,11 @@ to completeness or accuracy and it contains some references to files that are
 not part of the distribution.
 ================================================================================
 
+2023-12-06  Yukai Chou <muzimuzhi@gmail.com>
+	* ltcmd.dtx (subsection{Normalizing the argument specifications})
+	Use a clarified error message when an argument prefix `!' is applied to
```
Comment (josephwright) [inline base/changes.txt:11]: I've approved in review, but 'Clarify error message when ...' would be a little more natural

Comment (muzimuzhi) [reply base/changes.txt:11]: Oh I forgot to update all these places. Thanks.

**Review by FrankMittelbach — COMMENTED** [review]
I think that this adds in some sense unnecessary processing and just changing the explanation slightly would do as well (as I suggesed elsewhere). But if the common feeling is that this way it is more helpful, I'm not against that version.

**Comment by josephwright** [comment]
@FrankMittelbach Sure but we are in an error branch here so performance shouldn't be an issue? I agree it would be reasonable to simply make the description more accurate as an alternative fix.

**Review by FrankMittelbach — APPROVED** [review]
true only when erroring anyway, I misread that.

File: base/ltcmd.dtx [hunk 2]
```diff
@@ -4747,7 +4754,14 @@
...
-\msg_new:nnnn { cmd } { invalid-bang }
+\msg_new:nnnn { cmd } { invalid-bang-before-mandatory }
+  { Invalid~argument~prefix~'!'~in~#1. }
+  {
+    The~prefix~'!'~is~only~allowed~for~trailing~optional~arguments.~
+    You~tried~to~apply~it~to~an~optional~argument~before~mandatory~'#2'.
+    \c_@@_ignore_def_tl
+  }
+\msg_new:nnnn { cmd } { invalid-bang-on-mandatory }
   { Invalid~argument~prefix~'!'~in~#1. }
```
Comment (muzimuzhi) [inline base/ltcmd.dtx:4765]: I kinda changed my mind.

These two messages share the same `<text>` (`#3` of `\msg_new:nnnn`) and only differ in `<more text>` (`#4` of `\msg_new:nnnn`). Thus they can be combined.

Comment (muzimuzhi) [reply base/ltcmd.dtx]: Done in https://github.com/latex3/latex2e/pull/1199/commits/6ce39c0b3a29d7712c4be08bc7f880d3edd4f6e8, along with rebasing on `develop`.

**Pull request merged**